src/privatesend-util.cpp

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) 2014-2017 The Dash Core developers
+// Distributed under the MIT/X11 software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+#include "privatesend-util.h"
+
+KeyHolder::KeyHolder(CWallet* pwallet)
+        :reserveKey(pwallet)
+{
+        reserveKey.GetReservedKey(pubKey, false);
+}
+
+void KeyHolder::KeepKey()
+{
+    reserveKey.KeepKey();
+}
+
+void KeyHolder::ReturnKey()
+{
+    reserveKey.ReturnKey();
+}
+
+CScript KeyHolder::GetScriptForDestination()
+{
+    return ::GetScriptForDestination(pubKey.GetID());
+}
+
+
+KeyHolderPtr KeyHolderStorage::AddKey(CWallet* pwallet)
+{
+    LogPrintf("PrivateSend - KeyHolderStorage -- AddKey\n");
+    auto key = std::make_shared<KeyHolder>(pwallet);
+    storage.push_back(key);
+    return key;
+}
+
+void KeyHolderStorage::KeepAll(){
+    for(auto key : storage) {
+        key->KeepKey();
+    }
+    storage.clear();
+}
+
+void KeyHolderStorage::ReturnAll()
+{
+    for(auto key : storage) {
+        key->ReturnKey();
+    }
+    storage.clear();
+}
+
+void KeyHolderStorage::ClearOnFailure()
```

Comment:
What's the point in obfuscating the actual behavior via `ClearOn<smth>` instead of calling `KeepAll` and `ReturnAll` directly?